Disable onBytesReceived callback invocation when nobody is listening.

aam opened the pull request:
## Description

Don't invoke image loading onBytesReceived callback if there is nobody listening for those events.

## Related Issues

Performance traces show unnecessary callbacks load on UI isolate leading to dropped frames.

## Tests

I added the following tests:
```
diff --git a/packages/flutter/test/foundation/consolidate_response_test.dart b/packages/flutter/test/foundation/consolidate_response_test.dart
--- a/packages/flutter/test/foundation/consolidate_response_test.dart
+++ b/packages/flutter/test/foundation/consolidate_response_test.dart
+    test('Doesn\'t invoke onBytesReceived when controller is off', () async {
+    test('Invokes onBytesReceived when controller is on', () async {
+    test('Stops invoking onBytesReceived when controller turns off', () async {

```

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

aam committed:
Disable onBytesReceived callback invocation when nobody is listening.

tvolkert commented:
> I added the following tests:
>
> Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See Test Coverage.

Tests?

aam committed:
Add tests.

aam commented:
> > I added the following tests:
> > Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See Test Coverage.
> 
> Tests?

Added some now.

aam committed:
Add another test

goderbauer commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -47,6 +50,7 @@ Future<Uint8List> consolidateHttpClientResponseBytes(
   HttpClientResponse response, {
   bool autoUncompress = true,
   BytesReceivedCallback onBytesReceived,
+  SendChunkEventsController sendChunkEventsController,
```
This seems like an awkward API to have a callback and a separate object that controls if the callback gets called or not. I wonder if we can come up with something better here?

tvolkert replied:
+1.  This is not something that should be exposed in a public API.

aam replied:
Can you clarify your thought process on why this should not be exposed?
Proposed api is pretty straightforward: callback is invoked only when controller allows it do so. Potentially this can be wrapped into a single class (combining callback and a controller), but in current form it keeps compatibility with old api and reduces complexity of introducing another class.

tvolkert replied:
It's exposing an implementation detail.  The caller is already able to specify `onBytesReceived` or leave it null - it's not at all clear from an API perspective why they should also have to specify this opaque controller object.  Even if it was wrapped into a common class with one argument, the same would apply.

aam replied:
The fact that `onBytesReceived` is specified doesn't necessary mean that callback has to be invoked(https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/_network_image_io.dart#L94). It's only when user used `addListener` on listener produced by `load`(https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/_network_image_io.dart#L46) with non-null `onChunk` listener, then callback has to be invoked.

This is what this controller enables `NetworkImage` class that uses `consolidateHttpClientResponseBytes` to do.

goderbauer replied:
`consolidateHttpClientResponseBytes` itself is part of the public API, though, and should have a sensible API as well. As tvolkert said, just from looking at the API of this function its very odd that I have to specify this opaque controller object.

tvolkert replied:
Also, from an API perspective, `onBytesReceived` being specified *should* mean that it has to be invoked.  Perhaps we can fix the problem you're trying to solve inside the impl of `_loadAsync()`?

aam replied:
I don't immediately see how this can be solved in `_loadAsync`, if you have some particular insight please let me know.
Conceptually what I'm trying to express is that `onBytesReceived` provided by `_loadAsync` is _optional_, and have to be invoked only if user of `ImageStreamCompleter`(`ImageStreamCompleter` encapsulates `_loadAsync`) did register a callback _with_ `onChunk` handler.

goderbauer replied:
I also don't see an immediate solution. Maybe this requires a redesign to make this more sensible?

goderbauer commented on packages/flutter/lib/src/painting/_network_image_io.dart:
```diff
@@ -122,3 +126,37 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
+    if (listener.onChunk != null) {
+      chunkListeners.remove(listener);
+
+      if (chunkListeners.isEmpty) {
```
Could this use the super-classes `hasListeners` instead of separately keeping track of the listeners again?

aam replied:
no, it can't because here we are interested in those listeners that have `onChunk` specified.

goderbauer commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -127,3 +135,19 @@ class _OutputBuffer extends ByteConversionSinkBase {
+///  Encapsulates [isSomebodyListening] flag allowing for image loader
+///  to dynamically enable or disable sending of chunk events.
+class SendChunkEventsController {
+
```
nit: remove extra blank line.

goderbauer commented:
What is the performance bottleneck? Is it calling `onBytesReceived` or is it adding a `ImageChunkEvent` to a StreamController that nobody is listening to?

aam commented:
> What is the performance bottleneck? Is it calling `onBytesReceived` or is it adding a `ImageChunkEvent` to a StreamController that nobody is listening to? And how big is the difference in performance?

It's former - there are many `onBytesReceived` dart callbacks being invoked by Dart VM when nobody is actually listening. On prototypical "photos gallery" flutter app on its timeline it looks like avalanche of microscopic invocation of some dart code during  image loading, when the device is potentially busy with receiving the bytes, decoding already downloaded bytes, building frames with already decoded bytes.

The effect though might look more dramatic than it really is because those no-op callbacks even though invoked, don't interrupt or block frame construction(in other words, they won't cause jank).

goderbauer commented:
> The effect though might look more dramatic than it really is because those no-op callbacks even though invoked, don't interrupt or block frame construction(in other words, they won't cause jank).

Is it worthwhile to fix this problem then if it doesn't have much impact? Given our discussion above, fixing this probably requires a (major) redesign of how this works to get a sensible API...

aam commented:
> > The effect though might look more dramatic than it really is because those no-op callbacks even though invoked, don't interrupt or block frame construction(in other words, they won't cause jank).
> 
> Is it worthwhile to fix this problem then if it doesn't have much impact? Given our discussion above, fixing this probably requires a (major) redesign of how this works to get a sensible API...

Yeah, looks like it is not.
